File: packages/repo-gc-rust/src/reporting/terminal.rs
use colored::Colorize;
use crate::types::{Report, Severity};

pub fn render(report: &Report, no_color: bool) {
    if no_color {
        colored::control::set_override(false);
    }

    println!("\n{}", "═══ repo-gc — AI Context Efficiency ═══".bold());
    println!(
        "Files: {}  Lines: {}  Est. LLM tokens: ~{}k",
        report.files_analyzed,
        report.total_lines,
        report.total_estimated_tokens / 1000
    );
    if report.files_skipped > 0 {
        println!("  ({} files skipped — parse errors)\n", report.files_skipped);
    } else {
        println!();
    }

    let gs = &report.global_score;
    println!("{}", "Global Scores (0-100, higher = worse)".bold().underline());
    println!(
        "  AI Friction:        {}  {} ({})",
        score_bar(gs.ai_friction_score),
        colorize(gs.ai_friction_score),
        score_label(gs.ai_friction_score)
    );
    println!(
        "  Context Waste:      {}  {} ({})",
        score_bar(gs.context_waste_score),
        colorize(gs.context_waste_score),
        score_label(gs.context_waste_score)
    );
    println!(
        "  Structural Entropy: {}  {} ({})\n",
        score_bar(gs.structural_entropy_score),
        colorize(gs.structural_entropy_score),
        score_label(gs.structural_entropy_score)
    );

    println!(
        "{}  Your repo wastes ~{}% of agent context capacity (~{:.1}x Claude sessions)\n",
        "▸".yellow().bold(),
        gs.estimated_waste_pct,
        gs.context_waste_ratio
    );

    if report.findings.is_empty() {
        println!("{}", "No issues found — your codebase is AI-friendly.".green().bold());
        return;
    }

    let mut sorted = report.findings.clone();
    sorted.sort_by(|a, b| b.severity.weight().partial_cmp(&a.severity.weight()).unwrap());

    println!("{} ({} total)", "Detected Patterns".bold().underline(), sorted.len());
    for (i, f) in sorted.iter().enumerate() {
        let tok = f
            .estimated_tokens
            .map(|t| format!(" (~{}k tokens)", t / 1000))
            .unwrap_or_default();
        println!(
            "\n  {}. [{}] {} — {}{}",
            i + 1,
            sev_colored(&f.severity),
            f.path.display().to_string().cyan(),
            f.kind.label(),
            tok
        );
        for r in &f.reasons {
            println!("     • {}", r);
        }
    }
    println!();

    let top: Vec<_> = sorted
        .iter()
        .filter(|f| matches!(f.severity, Severity::Critical | Severity::High))
        .take(5)
        .collect();
    if !top.is_empty() {
        println!("{}", "Priority Cleanup Targets (reduce AI friction fastest)".bold().underline());
        for f in top {
            println!("  → {}", f.suggested_next_step);
        }
        println!();
    }
}

fn score_bar(score: u32) -> String {
    let filled = (score / 10).min(10) as usize;
    let empty = 10 - filled;
    format!("{}{}", "█".repeat(filled), "░".repeat(empty))
}

fn score_label(score: u32) -> &'static str {
    if score >= 70 {
        "HIGH"
    } else if score >= 40 {
        "MEDIUM"
    } else {
        "LOW"
    }
}

fn colorize(score: u32) -> String {
    let s = format!("{}/100", score);
    if score >= 70 {
        s.red().bold().to_string()
    } else if score >= 40 {
        s.yellow().to_string()
    } else {
        s.green().to_string()
    }
}

fn sev_colored(s: &Severity) -> colored::ColoredString {
    match s {
        Severity::Critical => s.label().red().bold(),
        Severity::High => s.label().red(),
        Severity::Medium => s.label().yellow(),
        Severity::Low => s.label().normal(),
    }
}

File: packages/repo-gc-python/python/repo_gc_python/reporters/terminal.py
"""Terminal reporter with ANSI color — mirrors Rust's terminal.rs."""

from ..types import Finding, FindingKind, Report, Severity

# ANSI escapes
BOLD = "\033[1m"
RED = "\033[31m"
YELLOW = "\033[33m"
GREEN = "\033[32m"
CYAN = "\033[36m"
MAGENTA = "\033[35m"
RESET = "\033[0m"
UNDERLINE = "\033[4m"


def _color(s: str, code: str) -> str:
    return f"{code}{s}{RESET}"


def _score_bar(score: int) -> str:
    filled = min(score // 10, 10)
    empty = 10 - filled
    return "█" * filled + "░" * empty


def _score_label(score: int) -> str:
    if score >= 70:
        return "HIGH"
    elif score >= 40:
        return "MEDIUM"
    else:
        return "LOW"


def _colorize(score: int) -> str:
    s = f"{score}/100"
    if score >= 70:
        return _color(s, RED + BOLD)
    elif score >= 40:
        return _color(s, YELLOW)
    else:
        return _color(s, GREEN)


def _sev_colored(s: Severity) -> str:
    if s == Severity.Critical:
        return _color(s.label, RED + BOLD)
    elif s == Severity.High:
        return _color(s.label, RED)
    elif s == Severity.Medium:
        return _color(s.label, YELLOW)
    else:
        return s.label


def render_terminal(report: Report, no_color: bool = False) -> str:
    use_color = not no_color
    c = lambda s, code: _color(s, code) if use_color else s

    lines: list[str] = []
    lines.append("")
    lines.append(c("═══ repo-gc — AI Context Efficiency ═══", BOLD))
    lines.append(
        f"Files: {report.files_analyzed}  Lines: {report.total_lines}  "
        f"Est. LLM tokens: ~{report.total_estimated_tokens // 1000}k"
    )
    if report.files_skipped > 0:
        lines.append(f"  ({report.files_skipped} files skipped — parse errors)\n")
    else:
        lines.append("")

    gs = report.global_score
    lines.append(c("Global Scores (0-100, higher = worse)", BOLD + UNDERLINE))
    lines.append(
        f"  AI Friction:        {_score_bar(gs.ai_friction_score)}  "
        f"{c(_colorize(gs.ai_friction_score), '') if use_color else _colorize(gs.ai_friction_score)} "
        f"({_score_label(gs.ai_friction_score)})"
    )
    lines.append(
        f"  Context Waste:      {_score_bar(gs.context_waste_score)}  "
        f"{c(_colorize(gs.context_waste_score), '') if use_color else _colorize(gs.context_waste_score)} "
        f"({_score_label(gs.context_waste_score)})"
    )
    lines.append(
        f"  Structural Entropy: {_score_bar(gs.structural_entropy_score)}  "
        f"{c(_colorize(gs.structural_entropy_score), '') if use_color else _colorize(gs.structural_entropy_score)} "
        f"({_score_label(gs.structural_entropy_score)})"
    )
    lines.append("")

    lines.append(
        f"{c('▸', YELLOW + BOLD) if use_color else '▸'}  "
        f"Your repo wastes ~{gs.estimated_waste_pct}% of agent context capacity "
        f"(~{gs.context_waste_ratio:.1f}x Claude sessions)"
    )
    lines.append("")

    if not report.findings:
        lines.append(c("No issues found — your codebase is AI-friendly.", GREEN + BOLD))
        return "\n".join(lines) + "\n"

    sorted_findings = sorted(report.findings, key=lambda f: f.severity.weight, reverse=True)

    lines.append(c(f"Detected Patterns ({len(sorted_findings)} total)", BOLD + UNDERLINE))
    for i, f in enumerate(sorted_findings):
        tok = f" (~{f.estimated_tokens // 1000}k tokens)" if f.estimated_tokens else ""
        sev_display = c(_sev_colored(f.severity), "") if use_color else f.severity.label
        lines.append(
            f"\n  {i + 1}. [{sev_display}] "
            f"{c(f.path, CYAN) if use_color else f.path} — {f.kind.label}{tok}"
        )
        for r in f.reasons:
            lines.append(f"     • {r}")
    lines.append("")

    top = [f for f in sorted_findings if f.severity in (Severity.Critical, Severity.High)][:5]
    if top:
        lines.append(c("Priority Cleanup Targets (reduce AI friction fastest)", BOLD + UNDERLINE))
        for f in top:
            lines.append(f"  → {f.suggested_next_step}")
        lines.append("")

    return "\n".join(lines) + "\n"
